src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (states.length > 0) {
+      currentState = { ...states[states.length - 1] };
+    }
+
+    switch (action.type) {
+      case 'addProperties':
+        const data = action.extraData;
+
+        for (const property in data) {
+          currentState[property] = data[property];
+        }
+        break;
+
+      case 'removeProperties':
+        const removes = action.keysToRemove;
+
+        for (const key of removes) {
+          if (key in currentState) {
+            delete currentState[key];
+          }
+        }
```

Comment:
```suggestion
        for (const key of removes) {
          delete currentState[key];
        }
```

can be more simplify